ntoskrnl/io/pnpmgr/pnpenum.c

```diff
@@ -0,0 +1,299 @@
+/*
+ * PROJECT:     ReactOS Kernel
+ * LICENSE:     GPL-2.0-or-later (https://spdx.org/licenses/GPL-2.0-or-later)
+ * PURPOSE:     Device enumeration functions
+ * COPYRIGHT:   2018 Vadim Galyant
+ */
+
+#include <ntoskrnl.h>
+
+#define NDEBUG
+#include <debug.h>
+
+extern KSPIN_LOCK IopPnPSpinLock;
+extern LIST_ENTRY IopPnpEnumerationRequestList;
+extern KEVENT PiEnumerationFinished;
+
+// Request types for PIP_ENUM_REQUEST
+typedef enum _PIP_ENUM_TYPE
+{
+    PipEnumAddBootDevices,
+    PipEnumAssignResources,
+    PipEnumGetSetDeviceStatus,
+    PipEnumClearProblem,
+    PipEnumInvalidateRelationsInList,
+    PipEnumHaltDevice,
+    PipEnumBootDevices,
+    PipEnumDeviceOnly,
+    PipEnumDeviceTree,
+    PipEnumRootDevices,
+    PipEnumInvalidateDeviceState,
+    PipEnumResetDevice,
+    PipEnumIoResourceChanged,
+    PipEnumSystemHiveLimitChange,
+    PipEnumSetProblem,
+    PipEnumShutdownPnpDevices,
+    PipEnumStartDevice,
+    PipEnumStartSystemDevices
+} PIP_ENUM_TYPE;
+
+typedef struct _PIP_ENUM_REQUEST
+{
+    LIST_ENTRY RequestLink;
+    PDEVICE_OBJECT DeviceObject;
+    PIP_ENUM_TYPE RequestType;
+    UCHAR ReorderingBarrier;
+    ULONG_PTR RequestArgument;
+    PKEVENT CompletionEvent;
+    NTSTATUS *CompletionStatus;
+} PIP_ENUM_REQUEST, *PPIP_ENUM_REQUEST;
+
+BOOLEAN PipEnumerationInProgress;
+WORK_QUEUE_ITEM PipDeviceEnumerationWorkItem;
+
+
+VOID
+NTAPI
+PipEnumerationWorker(_In_ PVOID Context)
```

Comment:
The previous function looks much cleaner than this. I suggest to get the code back and add a new PipSynchronousInvalidateDeviceRelations or something containing the body of this thing.